src/Formatter/JunitFormatter.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Roave\BackwardCompatibility\Formatter;
+
+use Psl\Str;
+use Roave\BackwardCompatibility\Changes;
+use Roave\BackwardCompatibility\Git\CheckedOutRepository;
+use Symfony\Component\Console\Output\OutputInterface;
+
+use function count;
+use function htmlspecialchars;
+use function implode;
+use function sprintf;
+use function trim;
+
+use const ENT_COMPAT;
+use const ENT_XML1;
+
+class JunitFormatter implements OutputFormatter
+{
+    public function __construct(
+        private OutputInterface $output,
+        private CheckedOutRepository $basePath,
+    ) {
+    }
+
+    public function write(Changes $changes): void
+    {
+        $basePath = $this->basePath->__toString() . '/';
+
+        $changeCount = count($changes);
+
+        $testcases = [];
+
+        foreach ($changes as $change) {
+            $filename = $change->file === null ? null : Str\replace($change->file, $basePath, '');
+
+            $name = implode(':', [
+                $this->escape($filename ?? ''),
+                $this->escape((string) ($change->line ?? '')),
+                $this->escape((string) ($change->column ?? '')),
+            ]);
+
+            $testcases[] = sprintf(
+                '  <testcase name="%s"><failure type="error" message="%s"/></testcase>',
```

Comment:
If we use an XML format, we should probably use a DOM builder for this, no? :thinking: 

String concatenation doesn't seem very healthy, although I see that you escape things accordingly

Reply:
i considered using a DOMDocument to build it but that would require the `dom` extension which i wasn't sure on the stance that you'd take

i can rework it to use it and have an `extension_loaded` check at the start of the formatter and throw if it's not present?

Reply:
Given that the escaping is done well, I think we can live with the string concatenation approach: we just need to make sure we note that in a comment (on the class) for posterity :+1: